File: assignement3/one/src/java/AssignmentThree.java
import java.io.*;
import java.util.*;
import java.lang.*;

class AssignmentThree {
    private String dictionary = "../doc/dict.txt";
    private String phonenumbers = "../doc/tele.txt";
    private ArrayList<String> numbers = new ArrayList<String>();
    private String originalNum;
    private StringBuffer phonenumber = new StringBuffer();
    private HashMap<String, String> wordsByNumbers = 
                                                new HashMap<String, String>();
    private String[] combinations = {"e","jnq","rwx","dsy","ft","am","civ",
                                                            "bku","lop","ghz"};
    public AssignmentThree () {
        readFromFiles();
    }
    
    public ArrayList<String> getNumbers() {
        return numbers;
    }
    
    public void readFromFiles() {
        try {
            BufferedReader brWords = new BufferedReader( new FileReader
                                                               ( dictionary ) );
            while ( brWords.ready() ) {
                String line = brWords.readLine().toLowerCase();
                wordsByNumbers.put( line, ( matchWordToNumbers( line ) ) );
            }
            BufferedReader brNumbers = new BufferedReader( new FileReader
                                                             ( phonenumbers ) );
            while ( brNumbers.ready() ) {
                numbers.add( brNumbers.readLine() );
            }
            brWords.close();
            brNumbers.close();
        } catch ( IOException ie ) {
            ie.printStackTrace();
        } catch ( NullPointerException ne ) { 
            System.out.println( "Krasch! " );
            ne.printStackTrace();
        }
    }
    
    public String matchWordToNumbers( String word ) {
        String correspondingNumber = "";
        for( int i = 0; i < word.length(); i++ ) {
            for( int y = 0; y < combinations.length ; y++ ) {
                if( combinations[ y ].contains( String.valueOf
                                                        ( word.charAt( i ) ) ) )
                    correspondingNumber += Integer.toString( y );
            }
        }
        return correspondingNumber;
    }
    
    public void startEncoding( String num ) {
        originalNum = num;
        encodeNumber( num, "");
    }
    
    public void encodeNumber( String num, String result ) {
        if ( num.length() == 0 )
            System.out.println( originalNum+": "+result );
        String substring = "";
        for (int i = 1 ; i <= num.length(); i++ ) {
            substring = num.substring( 0, i );
            for ( String n : wordsByNumbers.keySet() ) {
                if( substring.equals( wordsByNumbers.get( n ) ) ) {
                    encodeNumber( num.replaceFirst( substring, "" ), result+n+" ");
                }
            }
        }
    }
    
    public static void main( String[] args ) {
        AssignmentThree app = new AssignmentThree();
        ArrayList<String> telephonenumbers = app.getNumbers();
        for( int i = 0; i < telephonenumbers.size(); i++ )
            app.startEncoding( telephonenumbers.get(i) );
    }
}

File: assignement3/one/src/ruby/AssignmentThree.rb
class AssignmentThree

    attr_accessor :numbers, :words, :combinations

    def initialize
        readNumbers
        readWords
    end
    
    def readNumbers
        @numbers = []
        f = File.open( "../doc/tele.txt", "r" )
        f.each do | line |
            @numbers << line.strip.to_s
        end
    end
    
    def readWords
        @words = {}
        f = File.open( "../doc/dict.txt", "r" )
        f.each do | line |
            @words[ line.strip.downcase ] = matchWordToNumber( line.downcase )
        end
    end
    
    def matchWordToNumber( word )
        current = ""
        @combinations = ["e", "jnq", "rwx", "dsy", "ft", "am", "civ", "bku", 
                                                                  "lop", "ghz"]
        word.each_char do | char |
            @combinations.each do | s |
                if( s.include? char )
                    current += @combinations.index( s ).to_s
                end
            end
        end
        return current
    end
    
    def startEncoding
        @numbers.each do | number |
            encodeNumber( number, "", number )
        end
    end
    
    def encodeNumber( num, result, original )
        if( num.size == 0 )
            puts "#{original}: #{result}"
        end
        x = 1
        substring = ""
        num.size.times do
            substring = num.slice( 0, x )
            @words.each_pair do | key ,value |
                if( substring == value )
                    encodeNumber( num.sub( substring, "" ), result+key+" ", original )
                end
            end
            x += 1
        end
    end
end
